src/managed/Microsoft.Extensions.DependencyModel/DependencyContextJsonReader.Netstandard2_0.cs

```diff
@@ -0,0 +1,841 @@
+                        break;
+                    case DependencyContextStrings.TargetsPropertyName:
+                        targets = ReadTargets(ref jsonReader);
+                        break;
+                    case DependencyContextStrings.LibrariesPropertyName:
+                        libraryStubs = ReadLibraries(ref jsonReader);
+                        break;
+                    case DependencyContextStrings.RuntimesPropertyName:
+                        runtimeFallbacks = ReadRuntimes(ref jsonReader);
+                        break;
+                    default:
+                        jsonReader.Skip();
+                        break;
+                }
+            }
+
+            if (compilationOptions == null)
+            {
+                compilationOptions = CompilationOptions.Default;
+            }
+
+            Target runtimeTarget = SelectRuntimeTarget(targets, runtimeTargetName);
+            runtimeTargetName = runtimeTarget?.Name;
+
+            if (runtimeTargetName != null)
+            {
+                var seperatorIndex = runtimeTargetName.IndexOf(DependencyContextStrings.VersionSeperator);
+                if (seperatorIndex > -1 && seperatorIndex < runtimeTargetName.Length)
+                {
+                    runtime = runtimeTargetName.Substring(seperatorIndex + 1);
+                    framework = runtimeTargetName.Substring(0, seperatorIndex);
+                    isPortable = false;
+                }
+                else
+                {
+                    framework = runtimeTargetName;
+                }
+            }
+
+            Target compileTarget = null;
+
+            var ridlessTarget = targets.FirstOrDefault(t => !IsRuntimeTarget(t.Name));
+            if (ridlessTarget != null)
+            {
+                compileTarget = ridlessTarget;
+                if (runtimeTarget == null)
+                {
+                    runtimeTarget = compileTarget;
+                    framework = ridlessTarget.Name;
+                }
+            }
+
+            if (runtimeTarget == null)
+            {
+                throw new FormatException("No runtime target found");
+            }
+
+            // Holds document content, clear it before returning it.
+            rentedBuffer.AsSpan().Clear();
+            ArrayPool<byte>.Shared.Return(rentedBuffer);
+
+            return new DependencyContext(
+                new TargetInfo(framework, runtime, runtimeSignature, isPortable),
+                compilationOptions,
+                CreateLibraries(compileTarget?.Libraries, false, libraryStubs).Cast<CompilationLibrary>().ToArray(),
+                CreateLibraries(runtimeTarget.Libraries, true, libraryStubs).Cast<RuntimeLibrary>().ToArray(),
+                runtimeFallbacks ?? Enumerable.Empty<RuntimeFallbacks>());
+        }
+
+        private Target SelectRuntimeTarget(List<Target> targets, string runtimeTargetName)
```

Comment:
If the two different readers need to be maintained the code that's just copied between them should be in a shared file (essentially something like DependencyContextJsonReader.cs, DependencyContextJsonReader.SystemTextJson.cs, DependencyContextJsonReader.NewtonSoftJson.cs; where only the parts that are touching the JSON are in the satellite partials)

Reply:
Though, since it looks like you added extension methods to make for a fewer-lines-of-code-change, I wonder if the better answer for this project isn't to make project local reader type which just shims across the two other readers (it could be that there's a tricky thing somewhere, but flipping tabs shows that these are remarkably similar files)

Reply:
Here is the diff between the two implementations: https://www.diffchecker.com/TBfQTDQ0
Essentially, we have the following key differences:
- pass by ref explicitly (with the ref keyword) vs. pass by ref implicitly (passing a class)
- `(string)reader.Value` vs. `reader.GetString()`

We could consider doing if/defs around the function signatures OR pass the `JsonTextReader` class by ref explicitly (even though that is redundant). And then we could have a GetStringValue helper (similar to `IsTokenTypeProperty`). That would get rid of most of the diffs. Thoughts on this approach?

How could we shim across the readers?